lib/builddata.sh

```diff
@@ -1,29 +1,182 @@
-  else
-    meta_set "package_manager" "npm"
-  fi
-
-  meta_set "has_procfile" "$(if test -f "$BUILD_DIR/Procfile"; then echo "true"; else echo "false"; fi)"
-  meta_set "start_script" "$(read_json "$BUILD_DIR/package.json" ".scripts[\"start\"]")"
-  meta_set "postinstall_script" "$(read_json "$BUILD_DIR/package.json" ".scripts[\"postinstall\"]")"
-
-  meta_set "heroku_prebuild_script" "$(read_json "$BUILD_DIR/package.json" ".scripts[\"heroku-prebuild\"]")"
-  meta_set "build_script" "$(read_json "$BUILD_DIR/package.json" ".scripts[\"build\"]")"
-  meta_set "heroku_postbuild_script" "$(read_json "$BUILD_DIR/package.json" ".scripts[\"heroku-postbuild\"]")"
-  meta_set "heroku_cleanup_script" "$(read_json "$BUILD_DIR/package.json" ".scripts[\"heroku-cleanup\"]")"
-
-  # add any active features to the metadata set
-  # prefix the key with "feature-"
-  features_list | tr ' ' '\n' | while read -r key; do
-    if [[ -n $key ]]; then
-      meta_set "feature_$key" "$(features_get "$key")"
-    fi
-  done
+BUILD_DATA_FILE="${CACHE_DIR:?}/build-data/nodejs.json"
+PREVIOUS_BUILD_DATA_FILE="${CACHE_DIR:?}/build-data/nodejs-prev.json"
+
+# Legacy `key=value` format file used by older Node.js buildpack versions.
+LEGACY_BUILD_DATA_FILE="${CACHE_DIR:?}/build-data/nodejs"
+
+# Initializes the build data store, preserving the file from the previous build if it exists.
+# Call this at the start of `bin/compile` before using any other functions from this file.
+#
+# Usage:
+# ```
+# build_data::setup
+# ```
+function build_data::setup() {
+	if [[ -f "${BUILD_DATA_FILE}" ]]; then
+		# Rename the existing build data file rather than overwriting it, so we can lookup values
+		# from the previous build (such as when determining whether to invalidate the cache).
+		mv "${BUILD_DATA_FILE}" "${PREVIOUS_BUILD_DATA_FILE}"
+	else
+		mkdir -p "$(dirname "${BUILD_DATA_FILE}")"
+	fi
+
+	echo "{}" >"${BUILD_DATA_FILE}"
+}
+
+# Sets a string build data value. The value will be wrapped in double quotes and escaped for JSON.
+#
+# Usage:
+# ```
+# build_data::set_string "python_version" "1.2.3"
+# build_data::set_string "failure_reason" "install-dependencies::pip"
+# ```
+function build_data::set_string() {
+	local key="${1}"
+	local value="${2}"
+	build_data::_set "${key}" "${value}" "true"
+}
+
+# Sets a build data value for the elapsed time in seconds between the provided start time and the
+# current time, represented as a float with milliseconds precision.
+#
+# Usage:
+# ```
+# local dependencies_install_start_time=$(build_data::current_unix_realtime)
+# # ... some operation ...
+# build_data::set_duration "dependencies_install_duration" "${dependencies_install_start_time}"
+# ```
+function build_data::set_duration() {
+	local key="${1}"
+	local start_time="${2}"
+	local end_time duration
+	end_time="$(build_data::current_unix_realtime)"
+	duration="$(awk -v start="${start_time}" -v end="${end_time}" 'BEGIN { printf "%.3f", (end - start)/1000 }')"
```

Comment:
The value returned by `current_unix_realtime` has the extra precision stored as a float so we no longer need to divide by 1000 https://github.com/heroku/heroku-buildpack-python/commit/30086f2e8da2bf50d1b669b2e9ca82d479bf0892.

```suggestion
	duration="$(awk -v start="${start_time}" -v end="${end_time}" 'BEGIN { printf "%f", (end - start) }')"
```